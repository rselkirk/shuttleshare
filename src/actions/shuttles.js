import uuid from 'uuid';

// ADD_SHUTTLE
export const addShuttle = (
  { origin = '',
    destination = '',
    date = '',
    time = '',
    spots = '',
    cost = ''
  } = {}
) => ({
  type: 'ADD_SHUTTLE',
  shuttle: {
    id: uuid(),
    origin,
    destination,
    date,
    time,
    spots,
    cost
  }
});

// REMOVE_SHUTTLE
export const removeShuttle = ({ id } = {}) => ({
  type: 'REMOVE_SHUTTLE',
  id
});

// EDIT_SHUTTLE
export const editShuttle = (id, updates) => ({
  type: 'EDIT_SHUTTLE',
  id,
  updates
});