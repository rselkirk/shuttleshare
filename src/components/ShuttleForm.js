import React from 'react';
import PropTypes from 'prop-types';
import { withStyles } from '@material-ui/core/styles';
import InputLabel from '@material-ui/core/InputLabel';
import MenuItem from '@material-ui/core/MenuItem';
import FormControl from '@material-ui/core/FormControl';
import Select from '@material-ui/core/Select';
import Button from '@material-ui/core/Button';
import TextField from '@material-ui/core/TextField';

const styles = theme => ({
  root: {
    display: 'flex',
    flexWrap: 'wrap',
  },
  formControl: {
    margin: theme.spacing(1),
    minWidth: 120,
  },
  selectEmpty: {
    marginTop: theme.spacing(2),
  },
  button: {
    margin: theme.spacing(1),
  },
  textField: {
    marginLeft: theme.spacing(1),
    marginRight: theme.spacing(1),
    width: 200,
  },
});

class ShuttleForm extends React.Component {
  constructor(props) {
    super(props);

    this.state ={
      origin: props.shuttle ? props.shuttle.origin : '',
      destination: props.destination ? props.shuttle.destination : '',
      date: props.date ? props.shuttle.date : '',
      time: props.time ? props.shuttle.time : '',
      spots: props.spots ? props.shuttle.spots : '',
      cost: props.cost ? props.shuttle.cost : '',
      error: ''
    }
  }
  onOriginChange = (e) => {
    const origin = e.target.value;
    this.setState(() => ({ origin }));
  };
  onDestinationChange = (e) => {
    const destination = e.target.value;
    this.setState(() => ({ destination }));
  };
  onDateChange = (e) => {
    const date = e.target.value;
    this.setState(() => ({ date }));
  };
  onTimeChange = (e) => {
    const time = e.target.value;
    this.setState(() => ({ time }));
  };
  onSpotsChange = (e) => {
    const spots = e.target.value;
    this.setState(() => ({ spots }));
  };
  onCostChange = (e) => {
    const cost = e.target.value;
    this.setState(() => ({ cost }));
  };

  onSubmit = (e) => {
    e.preventDefault();

    if (!this.state.origin || !this.state.destination) {
      this.setState(() => ({ error: 'Please provide origin and destination.' }));
    } else {
      this.setState(() => ({ error: '' }));
      this.props.onSubmit({
        from: this.state.from,
        to: this.state.to,
        date: this.state.date,
        time: this.state.time,
        spots: e.target.elements.spots.value.trim(),
        cost: e.target.elements.cost.value.trim(),
      });
    }
  }
  
  render() {
    const { classes } = this.props;
    return (
      <div>
        {this.state.error && <p>{this.state.error}</p>}
        <form className={classes.root} onSubmit={this.onSubmit}>
          <FormControl className={classes.formControl}>
            <InputLabel htmlFor="origin">Origin</InputLabel>
            <Select
              onChange={this.onOriginChange}
              value={this.state.origin}
              inputProps={{
                name: 'origin'
              }}
            >
              <MenuItem value="Rossland">Rossland</MenuItem>
              <MenuItem value="Neptune Creek">Neptune Creek</MenuItem>
              <MenuItem value="Malde Creek">Malde Creek</MenuItem>
            </Select>
          </FormControl>
          <FormControl className={classes.formControl}>
            <InputLabel htmlFor="destination">Destination</InputLabel>
            <Select
              onChange={this.onDestinationChange}
              value={this.state.destination}
              inputProps={{
                name: 'destination'
              }}
            >
              <MenuItem value="Rossland">Rossland</MenuItem>
              <MenuItem value="Neptune Creek">Neptune Creek</MenuItem>
              <MenuItem value="Malde Creek">Malde Creek</MenuItem>
            </Select>
          </FormControl>
          <FormControl className={classes.formControl}>
            <TextField
              id="date"
              label="Departure Date"
              type="date"
              defaultValue="2019-09-24"
              className={classes.textField}
              InputLabelProps={{
                shrink: true,
              }}
            />
          </FormControl>
          <FormControl className={classes.formControl}>
            <TextField
              id="time"
              label="Departure Time"
              type="time"
              defaultValue="07:30"
              className={classes.textField}
              InputLabelProps={{
                shrink: true,
              }}
              inputProps={{
                step: 300, // 5 min
              }}
            />
          </FormControl>
          <FormControl className={classes.formControl}>
            <TextField
              className={classes.textField}
              onChange={this.onSpotsChange}
              value={this.state.spots}
              label="Spots"
              margin="normal"
              inputProps={{
                name: 'spots'
              }}
            />
          </FormControl>
          <FormControl className={classes.formControl}>
            <TextField
              className={classes.textField}
              onChange={this.onCostChange}
              value={this.state.cost}
              label="Cost"
              margin="normal"
              inputProps={{
                name: 'cost'
              }}
            />
          </FormControl>
          <Button variant="contained" className={classes.button} type="submit">Add Shuttle</Button>
          </form>
      </div>
    );
  }
}

ShuttleForm.propTypes = {
  classes: PropTypes.object.isRequired,
};

export default withStyles(styles)(ShuttleForm);