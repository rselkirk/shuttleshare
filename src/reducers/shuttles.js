// Shuttles reducer

const shuttlesReducerDefaultState = [];

export default (state = shuttlesReducerDefaultState, action) => {
  switch (action.type) {
    case 'ADD_SHUTTLE':
      console.log("shuttle added")
      console.log(action.shuttle);
      return [
        ...state,
        action.shuttle
      ];
    case 'REMOVE_SHUTTLE':
      return state.filter(({ id }) => id !== action.id);
    case 'EDIT_SHUTTLE':
      return state.map((shuttle) => {
        if (shuttle.id === action.id) {
          return {
            ...shuttle,
            ...action.updates
          };
        } else {
          return shuttle;
        };
      });
    default:
      return state;
  }
};
