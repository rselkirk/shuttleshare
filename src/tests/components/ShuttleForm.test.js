import React from 'react';
// import { shallow } from 'enzyme';
import { createShallow } from '@material-ui/core/test-utils';
import moment from 'moment';
import ShuttleForm from '../../components/ShuttleForm';
import shuttles from '../fixtures/shuttles';

let shallow;

beforeAll(() => {
  shallow = createShallow();
});

test('should render ShuttleForm correctly', () => {
  const wrapper = shallow(<ShuttleForm />);
  expect(wrapper).toMatchSnapshot();
});

test('should render ShuttleForm with shuttle data', () => {
  const wrapper = shallow(<ShuttleForm shuttle={shuttles[0]} />)
  expect(wrapper).toMatchSnapshot();
});

// test('should set new date on date change', () => {
// });

// test('should set calendar focus on change', () => {
// });
