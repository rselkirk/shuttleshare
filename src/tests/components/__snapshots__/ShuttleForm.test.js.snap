// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ShuttleForm correctly 1`] = `
<ShuttleForm
  classes={
    Object {
      "button": "ShuttleForm-button-4",
      "formControl": "ShuttleForm-formControl-2",
      "root": "ShuttleForm-root-1",
      "selectEmpty": "ShuttleForm-selectEmpty-3",
      "textField": "ShuttleForm-textField-5",
    }
  }
/>
`;

exports[`should render ShuttleForm with shuttle data 1`] = `
<ShuttleForm
  classes={
    Object {
      "button": "ShuttleForm-button-4",
      "formControl": "ShuttleForm-formControl-2",
      "root": "ShuttleForm-root-1",
      "selectEmpty": "ShuttleForm-selectEmpty-3",
      "textField": "ShuttleForm-textField-5",
    }
  }
  shuttle={
    Object {
      "cost": 500,
      "date": "1970-01-01T00:00:00.100Z",
      "destination": "Rossland",
      "id": 1,
      "origin": "Neptune Creek",
      "spots": 2,
      "time": "1970-01-01T00:00:00.100Z",
    }
  }
/>
`;
